### PR TITLE
Add design decisions to FAQ document

### PR DESCRIPTION
**Pull request by sbakkila**
I wrote up some answers for design decisions to the FAQ in response to [issue #1785](https://github.com/reactjs/redux/issues/1785)

I don't think that the md is perfect, and my page isn't showing up when I build the gitbook, but I wasn't sure how to proceed. Please advise on this. Thanks!

This is my first open source contribution. 😄

### CODE REVIEW

File: docs/faq/DesignDecisions.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+- [Why does applyMiddleware use a closure for dispatch?](#closure-dispatch)
+- [Can you please change combineReducers to support nested state trees?](#combineReducers-limitations)
+- [Why doesn't mapDispatchToProps allow use of return values from getState() or mapStateToProps()?](#no-asynch-in-mapDispatchToProps)
+
+
+## Design Decisions
+
+<a id="does-not-pass-state-action-to-subscribers"></a>
+### Why doesn't Redux pass the state and action to subscribers?
+Subscribers are intended to respond to the state value itself, not the action. Updates to the state are not always processed synchronously, because Redux sometimes processes actions in batches to optimize performance and repeated re-rendering. The intended guarantee is that Redux eventually calls all subscribers with the most recent state, but not that it always calls each subscriber for each action. The store state is available in the subscriber simply by calling store.getState(). The action cannot be made available in the subsribers without breaking the way that actions are batched.
```
Comment (timdorr) [inline docs/faq/DesignDecisions.md]: There is no batching in Redux natively. However, you can add that functionality. I might reword it to say "because libraries can change Redux to process actions in batches".

Comment (sbakkila) [inline docs/faq/DesignDecisions.md]: Thanks, good catch! Updated.

**Comment by sbakkila** [comment]
@markerikson @timdorr I just wanted to check again for any needed revisions. I have time to work on it this weekend if you have any suggestions! Thanks!

File: docs/faq/DesignDecisions.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+- [Why does applyMiddleware use a closure for dispatch?](#closure-dispatch)
+- [Can you please change combineReducers to support nested state trees?](#combineReducers-limitations)
+- [Why doesn't mapDispatchToProps allow use of return values from getState() or mapStateToProps()?](#no-asynch-in-mapDispatchToProps)
+
+
+## Design Decisions
+
+<a id="does-not-pass-state-action-to-subscribers"></a>
+### Why doesn't Redux pass the state and action to subscribers?
+Subscribers are intended to respond to the state value itself, not the action. Updates to the state are not always processed synchronously, because libraries can change Redux to process actions in batches to optimize performance and avoid repeated re-rendering. The intended guarantee is that Redux eventually calls all subscribers with the most recent state, but not that it always calls each subscriber for each action. The store state is available in the subscriber simply by calling store.getState(). The action cannot be made available in the subsribers without breaking the way that actions are batched.
```
Comment (markerikson) [inline docs/faq/DesignDecisions.md]: Also: add inline code backticks for `store.getState()`

File: docs/faq/DesignDecisions.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+<a id="does-not-support-classes"></a>
+### Why doesn't Redux support using classes for actions and reducers?
+The pattern of using functions, called action creators, to return action objects may seem counterintuitive to programmers with a lot of Object Oriented Programming experience, who would see this is a strong use-case for Classes and instances. Class instances for action objects and reducers are not supported because class instances make serialization and deserialization tricky. Deserialization methods like JSON.parse(string) will return a plain old Javascript object rather than class instances. 
+
+Serialization enables the brower to store all actions that have been dispatched, as well as the previous store states, with much less memory. Rewinding and 'hot reloading' the store is central to the Redux developer experience and the function of Redux DevTools. This also enables deserialized actions to be stored on the server and re-serialized in the brower in the case of server-side rendering with Redux.
```
Comment (markerikson) [inline docs/faq/DesignDecisions.md]: Add a link to the store FAQ entry on serializable values in the store.

File: docs/faq/DesignDecisions.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+2.) the [virtual DOM equality check (vDOMEq)](https://facebook.github.io/react/docs/optimizing-performance.html#avoid-reconciliation)
+3.) [React.PureComponent](https://facebook.github.io/react/docs/optimizing-performance.html#examples)
+4.) Using React-Redux: use [mapStateToProps](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options) to subscribe components to only the parts of the store that they need.
+
+<a id="does-not-support-classes"></a>
+### Why doesn't Redux support using classes for actions and reducers?
+The pattern of using functions, called action creators, to return action objects may seem counterintuitive to programmers with a lot of Object Oriented Programming experience, who would see this is a strong use-case for Classes and instances. Class instances for action objects and reducers are not supported because class instances make serialization and deserialization tricky. Deserialization methods like JSON.parse(string) will return a plain old Javascript object rather than class instances.
```
Comment (markerikson) [inline docs/faq/DesignDecisions.md]: General request: add backticks for inline code formatting to all mentions of function names and stuff like `JSON.parse()`.

File: docs/faq/DesignDecisions.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+<a id="why-currying"></a>
+### Why does the middleware signature use currying?
+The [curried function signature](https://github.com/reactjs/redux/issues/1744) of declaring middleware is [deemed unnecessary](https://github.com/reactjs/redux/pull/784) by some, because both store and next are available when the applyMiddleware function is executed. This issue has been determined to not be [worth introducing breaking changes](https://github.com/reactjs/redux/issues/1744).
+
+<a id="closure-dispatch"></a>
+### Why does applyMiddleware use a closure for dispatch?
+applyMiddleware takes the existing dispatch from the store and closes over it to create the initial chain of middlewares that have been invoked with an object that exposes the getState and dispatch functions, which enables middlewares that [rely on dispatch during initialization](https://github.com/reactjs/redux/pull/1592) to run. 
+
+<a id="combineReducers-limitations"></a>
+### Can you please change combineReducers to support nested state trees?
```
Comment (markerikson) [inline docs/faq/DesignDecisions.md]: Let's change the title on this one.  Maybe something like: "Why doesn't `combineReducers` include a third argument with the entire state?"

I'd also like to see the answer fleshed out a bit more.  Overall, the general answer is that `combineReducers` is deliberately opinionated, and it's also not immediately obvious what a potential third arg should be (entire state tree, some callback, some other part of the state tree, ... ?).  We _might_ change that in the future, but for now the answer is "if it doesn't fit your use case, you can write your own".

File: docs/faq/DesignDecisions.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+### Why doesn't mapDispatchToProps allow use of return values from getState() or mapStateToProps()?
+In general, connect generates a props object out of a closure that is injected with both the current state and dispatch. Asynchronous logic does not belong in the mapStateToProps and mapDispatchToProps functions at all. They should be only pure functions which transform the state to props and bind action creators to dispatch. 
+
+You cannot modify the state during the execution of mapStateToProps. Modifying the state from these functions could lead to infinite loops because every update would reinvoke the map functions. Calling getState inside mapStateToProps would always just return the same state that is passed to the function. 
+
+The preferred way to handle this use-case (needing to alter props based on the current state and mapDispatchToProps functions) is to work from  mergeProps, the third argument to the connect function. If specified, it is passed the result of mapStateToProps(), mapDispatchToProps(), and the container component's props. The plain object returned from mergeProps will be passed as props to the wrapped component.
+
+#### Further information
```
Comment (markerikson) [inline docs/faq/DesignDecisions.md]: I'd like these links moved to each be at the end of the appropriate question/answer.

File: docs/faq/DesignDecisions.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    }
+});
+```
+- reducers within combineReducers are only passed the part of the state that they modify, and not the top state.
+
+The default utility combineReducers is only one way to build a complex reducer. Consider using libraries like [combineSectionReducers](https://github.com/ryo33/combine-section-reducers) or [reduceReducers](https://github.com/acdlite/reduce-reducers) if you want your reducers to have a nested tree structure.
+
+<a id="no-asynch-in-mapDispatchToProps"></a>
+### Why doesn't mapDispatchToProps allow use of return values from getState() or mapStateToProps()?
+In general, connect generates a props object out of a closure that is injected with both the current state and dispatch. Asynchronous logic does not belong in the mapStateToProps and mapDispatchToProps functions at all. They should be only pure functions which transform the state to props and bind action creators to dispatch.
```
Comment (markerikson) [inline docs/faq/DesignDecisions.md]: Not sure I understand/follow the statements in these first two paragraphs.

I don't think we've really ever had anyone ask about calling `getState` inside of a `mapStateToProps` function.  The typical request/use case has been wanting to use either the entire state or the return value of `mapState` inside of `mapDispatch`, so that when functions are declared inside of `mapDispatch`, they can close over the latest returned values from the store.  We don't allow that approach in `mapDispatch` because it would mean also calling `mapDispatch` every time the store is updated, and also cause re-creation of functions every time, thus adding a lot of perf overhead.  

As the third paragraph says, if you _really_ want to do that, you can provide a `mergeProps` function and implement this behavior yourself, but we heavily discourage that approach.  We would much rather encourage use of component methods that pass props to bound action creators, like `this.props.toggleTodo(this.props.todoId)`.  (And, as my own personal preference, you shouldn't even write an actual `mapDispatch` function at all - instead, write separate action creator functions, and use the object shorthand form of `mapDispatch` to bind them automatically.)

Would appreciate a rework of this answer.

**Comment by markerikson** [comment]
@sbakkila : Thanks for working on this!  Left several requests for changes, but this is definitely on the right track.  Appreciate it!

**Comment by sbakkila** [comment]
@markerikson thanks for the feedback! I have made the suggested edits, and pushed changes to my fork. Let me know if you see anything else for me to fix.

**Comment by markerikson** [comment]
Tell you what.  I see a couple more nitpicky things I'd like to tweak, but I think this is sufficiently good for now, so I'll go ahead and merge it.  

Thank you VERY much for contributing this section, and working with me to improve it.  I still need to go back to the "FAQ Updates" thread and figure out what other stuff still needs to be added, but if you see anything else that's not in the FAQ yet, please feel free to start adding those too.
